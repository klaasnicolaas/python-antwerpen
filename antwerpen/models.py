"""Asynchronous Python client providing Open Data information of Antwerpen."""
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Any


@dataclass
class DisabledParking:
    """Object representing a disabled parking."""

    entry_id: int
    number: int
    color: str | None
    address: str
    gis_id: str
    created_at: datetime | None
    longitude: float
    latitude: float

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> DisabledParking:
        """Return a DisabledParking object from a dictionary.

        Args:
            data: The data from the API.

        Returns:
            A DisabledParking object.
        """

        attr = data["properties"]
        geo = data["geometry"]["coordinates"]
        return cls(
            entry_id=attr.get("OBJECTID"),
            number=attr.get("AANTAL_PLAATSEN"),
            color=attr.get("KLEUR"),
            address=attr.get("ADRES"),
            gis_id=attr.get("GISID"),
            created_at=fromtimestamp(attr.get("START_DATE")),
            longitude=geo[0],
            latitude=geo[1],
        )


def fromtimestamp(epoch_time: int, default: None = None) -> Any:
    """Fromtimestamp function with default value.

    Args:
        epoch_time: The time in epoch format.
        default: The default value.

    Returns:
        The datetime object.
    """
    try:
        return datetime.fromtimestamp(epoch_time / 1000)
    except (ValueError, TypeError):
        return default
