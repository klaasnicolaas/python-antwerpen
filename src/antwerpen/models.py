"""Asynchronous Python client providing Open Data information of Antwerpen."""
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Any

import pytz


@dataclass
class DisabledParking:
    """Object representing a disabled parking."""

    entry_id: int
    number: int
    orientation: str
    destiny: str
    window_time: str | None
    lined: bool
    status: str
    gis_id: str
    created_at: datetime | None
    coordinates: list[float]

    @classmethod
    def from_dict(cls: type[DisabledParking], data: dict[str, Any]) -> DisabledParking:
        """Return a DisabledParking object from a dictionary.

        Args:
        ----
            data: The data from the API.

        Returns:
        -------
            A DisabledParking object.
        """
        attr = data["properties"]
        return cls(
            entry_id=attr.get("OBJECTID"),
            number=attr.get("Capaciteit"),
            orientation=attr.get("Orientatie"),
            destiny=attr.get("Bestemming"),
            window_time=attr.get("VENSTERTIJD_BESCHR"),
            lined=attr.get("GELIJND") == "Ja",
            status=attr.get("STATUS"),
            gis_id=attr.get("GISID"),
            created_at=fromtimestamp(attr.get("EBDD")),
            coordinates=data["geometry"]["coordinates"][0],
        )


def fromtimestamp(epoch_time: int, default: None = None) -> Any:
    """Fromtimestamp function with default value.

    Args:
    ----
        epoch_time: The time in epoch format.
        default: The default value.

    Returns:
    -------
        The datetime object.
    """
    try:
        return datetime.fromtimestamp(
            epoch_time / 1000,
            tz=pytz.timezone("Europe/Brussels"),
        )
    except (ValueError, TypeError):
        return default
