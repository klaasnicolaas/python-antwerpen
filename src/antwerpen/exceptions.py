"""Asynchronous Python client providing Open Data information of Antwerpen."""


class ODPAntwerpenError(Exception):
    """Generic Open Data Platform Antwerpen exception."""


class ODPAntwerpenConnectionError(ODPAntwerpenError):
    """Open Data Platform Antwerpen - connection error."""
