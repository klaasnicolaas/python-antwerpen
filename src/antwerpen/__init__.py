"""Asynchronous Python client providing Open Data information of Antwerpen."""

from .antwerpen import ODPAntwerpen
from .exceptions import ODPAntwerpenConnectionError, ODPAntwerpenError
from .models import DisabledParking

__all__ = [
    "ODPAntwerpen",
    "ODPAntwerpenConnectionError",
    "ODPAntwerpenError",
    "DisabledParking",
]
