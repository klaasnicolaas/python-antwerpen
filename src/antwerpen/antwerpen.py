"""Asynchronous Python client providing Open Data information of Antwerpen."""
from __future__ import annotations

import asyncio
import socket
from dataclasses import dataclass
from importlib import metadata
from typing import Any

import async_timeout
from aiohttp import ClientError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from .exceptions import ODPAntwerpenConnectionError, ODPAntwerpenError
from .models import DisabledParking


@dataclass
class ODPAntwerpen:
    """Main class for handling data fetchting from Open Data Platform of Antwerpen."""

    request_timeout: float = 10.0
    session: ClientSession | None = None

    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Handle a request to the Open Data Platform API of Antwerpen.

        Args:
        ----
            uri: Request URI, without '/', for example, 'status'
            method: HTTP method to use, for example, 'GET'
            params: Extra options to improve or limit the response.

        Returns:
        -------
            A Python dictionary (text) with the response from
            the Open Data Platform API.

        Raises:
        ------
            ODPAntwerpenConnectionError: Timeout occurred while
                connecting to the Open Data Platform API.
            ODPAntwerpenError: If the data is not valid.
        """
        version = metadata.version(__package__)
        url = URL.build(
            scheme="https",
            host="geodata.antwerpen.be",
            path="/arcgissql/rest/services/P_Portal/",
        ).join(URL(uri))

        headers = {
            "Accept": "application/geo+json",
            "User-Agent": f"PythonODPAntwerpen/{version}",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    params=params,
                    headers=headers,
                    ssl=True,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            msg = "Timeout occurred while connecting to the Open Data Platform API."
            raise ODPAntwerpenConnectionError(msg) from exception
        except (ClientError, socket.gaierror) as exception:
            msg = "Error occurred while communicating with Open Data Platform API."
            raise ODPAntwerpenConnectionError(msg) from exception

        content_type = response.headers.get("Content-Type", "")
        if "application/geo+json" not in content_type:
            text = await response.text()
            msg = "Unexpected content type response from the Open Data Platform API."
            raise ODPAntwerpenError(
                msg,
                {"Content-Type": content_type, "Response": text},
            )

        return await response.json()

    async def disabled_parkings(self, limit: int = 10) -> list[DisabledParking]:
        """Get all disabled parking spaces.

        Args:
        ----
            limit: Number of items to return.

        Returns:
        -------
            A list of DisabledParking objects.
        """
        results: list[DisabledParking] = []
        for offset in [0, 1000]:
            locations = await self._request(
                "portal_publiek6/MapServer/585/query",
                params={
                    "where": "1=1",
                    "resultOffset": offset,
                    "outFields": "*",
                    "f": "geojson",
                },
            )
            for location in locations["features"]:
                results.append(DisabledParking.from_dict(location))  # noqa: PERF401
        return results[:limit]

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> ODPAntwerpen:
        """Async enter.

        Returns
        -------
            The Open Data Platform object.
        """
        return self

    async def __aexit__(self, *_exc_info: str) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.
        """
        await self.close()
