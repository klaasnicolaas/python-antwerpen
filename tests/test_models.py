"""Test the models."""
from aiohttp import ClientSession
from aresponses import ResponsesMockServer

from antwerpen import DisabledParking, ODPAntwerpen

from . import load_fixtures


async def test_all_disabled_parking_spaces(aresponses: ResponsesMockServer) -> None:
    """Test all disabled parking spaces function."""
    aresponses.add(
        "geodata.antwerpen.be",
        "/arcgissql/rest/services/P_Portal/portal_publiek6/MapServer/585/query",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/geo+json"},
            text=load_fixtures("disabled_parking.geojson"),
        ),
    )
    async with ClientSession() as session:
        client = ODPAntwerpen(session=session)
        spaces: list[DisabledParking] = await client.disabled_parkings()
        assert spaces is not None
        for item in spaces:
            assert isinstance(item, DisabledParking)
            assert item.entry_id is not None
            assert item.coordinates is not None
            assert isinstance(item.coordinates, list)
