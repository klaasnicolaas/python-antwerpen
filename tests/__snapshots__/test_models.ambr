# serializer version: 1
# name: test_all_disabled_parking_spaces
  list([
    DisabledParking(entry_id=42044, number=1, orientation='Langs', destiny='mindervalide', window_time=None, lined=False, status='actief', gis_id='GIS_J_4213', created_at=datetime.datetime(2022, 7, 12, 2, 0, tzinfo=<DstTzInfo 'Europe/Brussels' CEST+2:00:00 DST>), coordinates=[[4.391823216523055, 51.204380414721534], [4.391890655441355, 51.20439166725936], [4.391920670914526, 51.20439667495787], [4.391912910140442, 51.204413466938526], [4.391912672866264, 51.20441394067745], [4.391815870951576, 51.204397789063655], [4.391823216523055, 51.204380414721534]]),
    DisabledParking(entry_id=42045, number=1, orientation='Langs', destiny='mindervalide', window_time=None, lined=False, status='actief', gis_id='GIS_102', created_at=datetime.datetime(2020, 2, 6, 1, 0, tzinfo=<DstTzInfo 'Europe/Brussels' CET+1:00:00 STD>), coordinates=[[4.455675140268877, 51.23493059018865], [4.455645047689535, 51.234920763051], [4.455691403116768, 51.234864822120784], [4.455721494237986, 51.23487464924792], [4.455675140268877, 51.23493059018865]]),
    DisabledParking(entry_id=42046, number=1, orientation='Langs', destiny='mindervalide', window_time=None, lined=False, status='actief', gis_id='GIS_103', created_at=datetime.datetime(2019, 12, 10, 1, 0, tzinfo=<DstTzInfo 'Europe/Brussels' CET+1:00:00 STD>), coordinates=[[4.426031684618144, 51.21583231636185], [4.426003855387162, 51.21584377404424], [4.425946830909598, 51.21579469741439], [4.4259733277135656, 51.21578209437603], [4.426031684618144, 51.21583231636185]]),
    DisabledParking(entry_id=42047, number=1, orientation='Langs', destiny='mindervalide', window_time=None, lined=False, status='actief', gis_id='GIS_109', created_at=datetime.datetime(2020, 1, 29, 1, 0, tzinfo=<DstTzInfo 'Europe/Brussels' CET+1:00:00 STD>), coordinates=[[4.4077970933008155, 51.2028035236245], [4.4078555624104, 51.20285450891529], [4.40783234560851, 51.20286501703832], [4.407774278858268, 51.20281438394837], [4.4077970933008155, 51.2028035236245]]),
    DisabledParking(entry_id=42048, number=1, orientation='Dwars', destiny='mindervalide', window_time=None, lined=False, status='actief', gis_id='GIS_110', created_at=datetime.datetime(2019, 8, 9, 2, 0, tzinfo=<DstTzInfo 'Europe/Brussels' CEST+2:00:00 DST>), coordinates=[[4.380429025818914, 51.2256858501889], [4.380349785386456, 51.22568696222369], [4.38034909565823, 51.22566616230958], [4.3803490538075245, 51.22566488057244], [4.38042838580686, 51.225663730777406], [4.380428417590236, 51.22566482735518], [4.380429025818914, 51.2256858501889]]),
    DisabledParking(entry_id=42049, number=1, orientation='Dwars', destiny='mindervalide', window_time=None, lined=False, status='actief', gis_id='GIS_111', created_at=datetime.datetime(2019, 8, 9, 2, 0, tzinfo=<DstTzInfo 'Europe/Brussels' CEST+2:00:00 DST>), coordinates=[[4.380429665828582, 51.22570795791531], [4.380350519835611, 51.22570906904189], [4.380349785386456, 51.22568696222369], [4.380429025818914, 51.2256858501889], [4.380429665828582, 51.22570795791531]]),
    DisabledParking(entry_id=42050, number=1, orientation='Langs', destiny='mindervalide', window_time=None, lined=False, status='actief', gis_id='GIS_113', created_at=datetime.datetime(2020, 2, 7, 1, 0, tzinfo=<DstTzInfo 'Europe/Brussels' CET+1:00:00 STD>), coordinates=[[4.449146791285996, 51.22905234261221], [4.449212533133099, 51.229007611870834], [4.449233798323922, 51.22901942955687], [4.449168083649293, 51.22906414231456], [4.449146791285996, 51.22905234261221]]),
    DisabledParking(entry_id=42051, number=1, orientation='Langs', destiny='mindervalide', window_time=None, lined=False, status='actief', gis_id='GIS_1132', created_at=datetime.datetime(2019, 9, 30, 2, 0, tzinfo=<DstTzInfo 'Europe/Brussels' CEST+2:00:00 DST>), coordinates=[[4.419602987720314, 51.228704999836886], [4.419627146201948, 51.228695344793614], [4.419680052663834, 51.22874965994693], [4.419655629020243, 51.22875904007267], [4.419655318016714, 51.22875872112155], [4.419602987720314, 51.228704999836886]]),
    DisabledParking(entry_id=42052, number=1, orientation='Langs', destiny='mindervalide', window_time=None, lined=False, status='actief', gis_id='GIS_1138', created_at=datetime.datetime(2019, 9, 30, 2, 0, tzinfo=<DstTzInfo 'Europe/Brussels' CEST+2:00:00 DST>), coordinates=[[4.420449466563296, 51.229550742572265], [4.420473592352026, 51.22954106130764], [4.420515977169261, 51.22958271254125], [4.420491835603105, 51.22959237854114], [4.420449466563296, 51.229550742572265]]),
    DisabledParking(entry_id=42053, number=1, orientation='Langs', destiny='mindervalide', window_time=None, lined=False, status='actief', gis_id='GIS_1139', created_at=None, coordinates=[[4.420499478750253, 51.22959989031804], [4.4204918370347235, 51.22959237854052], [4.420515977169261, 51.22958271254125], [4.420523786566688, 51.22959038693213], [4.420558223045742, 51.22962422899532], [4.42053409436538, 51.229633907581686], [4.420499478750253, 51.22959989031804]]),
  ])
# ---
